packages/flutter/test/widgets/scrollable_test.dart

```diff
@@ -676,4 +676,144 @@ void main() {
+        itemBuilder: (BuildContext context, int index) {
+          if (Scrollable.recommendDeferredLoadingForContext(context)) {
+            cheapWidgets += 1;
+            return const SizedBox(height: 50.0);
+          }
+          expensiveWidgets += 1;
+          return SizedBox(key: ValueKey<String>('Box $index'), height: 50.0);
+        },
+      ),
+    ));
+
+    await tester.pumpAndSettle();
+    expect(find.byKey(const ValueKey<String>('Box 0')), findsOneWidget);
+    expect(find.byKey(const ValueKey<String>('Box 111')), findsNothing);
+
+    expect(expensiveWidgets, 17);
+    expect(cheapWidgets, 0);
+
+    // The position value here is different from the maximum velocity we will
+    // reach, which is controlled by a combination of curve, duration, and
+    // position.
+    // This is just meant to be a pretty good simulation. A linear curve
+    // with these same parameters will never back off on the velocity enough
+    // to reset here.
+    controller.animateTo(
+      5000,
+      duration: const Duration(seconds: 2),
+      curve: Curves.fastLinearToSlowEaseIn,
+    );
+
+    await tester.pumpAndSettle();
+    expect(find.byKey(const ValueKey<String>('Box 0')), findsNothing);
+    expect(find.byKey(const ValueKey<String>('Box 111')), findsOneWidget);
+
+    expect(expensiveWidgets, 69);
+    expect(cheapWidgets, 48);
+  });
+
+  testWidgets('Can recommendDeferredLoadingForContext - ballistics', (WidgetTester tester) async {
+    int cheapWidgets = 0;
+    int expensiveWidgets = 0;
+    await tester.pumpWidget(Directionality(
+      textDirection: TextDirection.ltr,
+      child: ListView.builder(
+        physics: const NeverScrollableScrollPhysics(),
+        itemBuilder: (BuildContext context, int index) {
+          if (Scrollable.recommendDeferredLoadingForContext(context)) {
+            cheapWidgets += 1;
+            return const SizedBox(height: 50.0);
+          }
+          expensiveWidgets += 1;
+          return SizedBox(key: ValueKey<String>('Box $index'), height: 50.0);
+        },
+      ),
+    ));
+
+    await tester.pumpAndSettle();
+    expect(find.byKey(const ValueKey<String>('Box 0')), findsOneWidget);
+    expect(find.byKey(const ValueKey<String>('Box 52')), findsNothing);
+
+    expect(expensiveWidgets, 17);
+    expect(cheapWidgets, 0);
+
+    // Getting the tester to simulate a life-like fling is difficult.
+    // Instead, just manually drive the activity with a ballistic simulation as
+    // if the user has flung the list.
+    Scrollable.of(find.byType(SizedBox).evaluate().first).position.activity.delegate.goBallistic(4000);
+
+    await tester.pumpAndSettle();
+    expect(find.byKey(const ValueKey<String>('Box 0')), findsNothing);
+    expect(find.byKey(const ValueKey<String>('Box 52')), findsOneWidget);
+
+    expect(expensiveWidgets, 38);
+    expect(cheapWidgets, 20);
+  });
+
+  testWidgets('Can recommendDeferredLoadingForContext - override heuristic', (WidgetTester tester) async {
```

Comment:
how does this test show that the overridden method was actually used?

Maybe (also) have a test with physics that always defer no mater what and check that it only builds cheap boxes?

Reply:
I'll add state to the physics to assert that it's called.

I think with that the other test isn't strictly necessary, right?

Reply:
I guess... although it could still be a wired combination of some kind of default defer logic and this :P

Reply:
Ahh fair.  Ok.